### PR TITLE
media-fonts/Shanggu: new package, add 1.021

### LINKED ISSUE #5648: Additional Package Request: media-fonts/shanggu
Shanggu is an [open source](https://github.com/GuiWonder/Shanggu/blob/main/LICENSE.txt) CJK font that uses traditional letterforms; (_jiù zìxíng_/_kyūjitai_) a fork of [Noto CJK](https://github.com/notofonts/noto-cjk).

Upstream URL: https://github.com/GuiWonder/Shanggu

---

I have tried to create my own ebuild, but it is archived in 7zip and cannot be expanded.

How can I extract 7zip with ebuild?

### PR DESCRIPTION
**Pull request by Linerre**
fix #5648 

@curlingchimpanzeeowl if you'd like to maintain this font, feel free to modify the `overlay.toml` file：
```
- github_account = "Linerre"` 
+ github_account = "curlingchimpanzeeowl"
```
so that you will be notified when there is a newer version of this font released by the upstream.

You may also want to change the maintainer in the `metadata.xml` file to be consistent.

I have tested other font families by issuing the following
```
USE="sans mono round" emerge -av1 media-fonts/Shanggu"
```

The full installation takes up 4.5GB however, just for your reference:
```
[ebuild  N     ] media-fonts/Shanggu-1.021::gentoo-zh  USE="X mono round sans serif" 0 KiB

Total: 1 package (1 new), Size of downloads: 0 KiB

Would you like to merge these packages? [Yes/No] 
>>> Verifying ebuild manifests
>>> Emerging (1 of 1) media-fonts/Shanggu-1.021::gentoo-zh
>>> Installing (1 of 1) media-fonts/Shanggu-1.021::gentoo-zh
>>> Completed (1 of 1) media-fonts/Shanggu-1.021::gentoo-zh
>>> Jobs: 1 of 1 complete                                               Load avg: 0.67, 0.24, 0.08

 * Messages for package media-fonts/Shanggu-1.021:
 * Log file: /var/log/portage/build/media-fonts/Shanggu-1.021:20241117-041516.log

 * Package:    media-fonts/Shanggu-1.021:0
 * Repository: gentoo-zh
 * USE:        X abi_x86_64 amd64 elibc_glibc kernel_linux mono round sans serif
 * FEATURES:   network-sandbox preserve-libs sandbox userpriv usersandbox
 * Creating fonts.scale & fonts.dir in ShangguSerif
 * Creating fonts.scale & fonts.dir in ShangguSerifFANTI
 * Creating fonts.scale & fonts.dir in ShangguSerifJP
 * Creating fonts.scale & fonts.dir in ShangguSerifSC
 * Creating fonts.scale & fonts.dir in ShangguSerifTC
 * Creating fonts.scale & fonts.dir in ShangguSans
 * Creating fonts.scale & fonts.dir in ShangguSansFANTI
 * Creating fonts.scale & fonts.dir in ShangguSansJP
 * Creating fonts.scale & fonts.dir in ShangguSansSC
 * Creating fonts.scale & fonts.dir in ShangguSansTC
 * Creating fonts.scale & fonts.dir in ShangguMono
 * Creating fonts.scale & fonts.dir in ShangguMonoJP
 * Creating fonts.scale & fonts.dir in ShangguMonoSC
 * Creating fonts.scale & fonts.dir in ShangguMonoTC
 * Creating fonts.scale & fonts.dir in ShangguRound
 * Creating fonts.scale & fonts.dir in ShangguRoundFANTI
 * Creating fonts.scale & fonts.dir in ShangguRoundJP
 * Creating fonts.scale & fonts.dir in ShangguRoundSC
 * Creating fonts.scale & fonts.dir in ShangguRoundTC
 * Final size of build directory: 4815880 KiB (4.5 GiB)
 * Final size of installed tree:  4816556 KiB (4.5 GiB)
 * Updating global fontcache ...

 * GNU info directory index is up-to-date.
```

### CODE REVIEW

File: media-fonts/Shanggu/Shanggu-1.021.ebuild [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+src_unpack() {
+	use mono && unpack_7z "${PN}Mono-${PV}.7z"
+	use round && unpack_7z "${PN}Round-${PV}.7z"
+	use sans && unpack_7z "${PN}Sans-${PV}.7z"
+	use serif && unpack_7z "${PN}Serif-${PV}.7z"
+}
+
+src_install() {
+	if use serif; then
+		FONT_S=(
```
Comment (liangyongxiang) [inline media-fonts/Shanggu/Shanggu-1.021.ebuild:39]: 是否可以用数组append的方式更新FONT_S，这样整个src_install都可以省略。
https://stackoverflow.com/questions/1951506/add-a-new-element-to-an-array-without-specifying-the-index-in-bash

Comment (Linerre) [reply media-fonts/Shanggu/Shanggu-1.021.ebuild:39]: 我看看能不能把显式的 `src_install` 去掉

Comment (Linerre) [reply media-fonts/Shanggu/Shanggu-1.021.ebuild:39]:
```suggestion
FONT_S=()
use serif && FONT_S+=(
    "${WORKDIR}/${PN}Serif"
    "${WORKDIR}/${PN}SerifFANTI"
    "${WORKDIR}/${PN}SerifJP"
    "${WORKDIR}/${PN}SerifSC"
    "${WORKDIR}/${PN}SerifTC"
)
use sans && FONT_S+=(
    "${WORKDIR}/${PN}Sans"
    "${WORKDIR}/${PN}SansFANTI"
    "${WORKDIR}/${PN}SansJP"
    "${WORKDIR}/${PN}SansSC"
    "${WORKDIR}/${PN}SansTC"
)
use mono && FONT_S+=(
    "${WORKDIR}/${PN}Mono"
    "${WORKDIR}/${PN}MonoJP"
    "${WORKDIR}/${PN}MonoSC"
    "${WORKDIR}/${PN}MonoTC"
)
use round && FONT_S+=(
    "${WORKDIR}/${PN}Round"
    "${WORKDIR}/${PN}RoundFANTI"
    "${WORKDIR}/${PN}RoundJP"
    "${WORKDIR}/${PN}RoundSC"
    "${WORKDIR}/${PN}RoundTC"
)
```
这样呢？ 但好像没见过 `use flagxxx` 在 phase function 外面用的……

Comment (liangyongxiang) [reply media-fonts/Shanggu/Shanggu-1.021.ebuild:39]: 抱歉，这几天比较忙，都忘记这个事情了。看起来的确不如第一版。（逃

Comment (liangyongxiang) [reply media-fonts/Shanggu/Shanggu-1.021.ebuild:39]: 应该是我没说清楚，我原来的想法是，if use sans; then FON_S+=(); fi 之后，最后调用一次 font_src_install 就可以了。

Comment (liangyongxiang) [reply media-fonts/Shanggu/Shanggu-1.021.ebuild:39]: 我先合并吧。

Comment (Linerre) [reply media-fonts/Shanggu/Shanggu-1.021.ebuild:39]: > 我原来的想法是，if use sans; then FON_S+=(); fi 之后，最后调用一次 font_src_install 就可以了。

我懂了，那我下次 bump 时改一下试试，多谢～

**Pull request merged**